pkg/didcomm/protocol/route/service.go

```diff
@@ -296,6 +328,80 @@ func (s *Service) handleForward(msg *service.DIDCommMsg) error {
 	return s.outbound.Forward(forward.Msg, dest)
 }
 
+// Register registers the agent with a router with the on connectionID passed in. The agent is
+// registered with the router and retrieves the router endpoint and routing keys. This function
+// throws an error if the agent is already registered against a router. This function internally
+// works as a sync function and works in-conjunction with handleGrant function. ie, The request
+// did-comm message is sent to the router and it waits for the grant response from the router
+// till timeout.
+func (s *Service) Register(connectionID string) error {
+	// check if router is already registered
+	routerConnID, err := s.getRouterConnectionID()
+	if err != nil && !errors.Is(err, storage.ErrDataNotFound) {
+		return fmt.Errorf("fetch router connection id : %w", err)
+	}
+
+	if routerConnID != "" {
+		return errors.New("router is already registered")
+	}
+
+	// get the connection record for the ID to fetch DID information
+	conn, err := s.connectionLookup.GetConnectionRecord(connectionID)
+	if err != nil {
```

Comment:
Any reasons for separating these data not found error? Any handling in the caller?

Reply:
yeah. we need to handle these cases differently in the caller ie. If router connection is present and that connection is not found, then user should be able to add a new router connectionID.